### ISSUE TITLE
Allow TLS hostname overrides

### ISSUE THREAD

**Issue by crivera-fastly** [confluentinc/librdkafka#3321]
Description
===========

It would be useful to have a way to override the hostname used for TLS hostname verification. For reference, the Go TLS stack provides a ServerName field for this purpose: [tls - The Go Programming Language](https://golang.org/pkg/crypto/tls/#Config). This would allow clients to specify a trusted name for scenarios that would otherwise require modifications to the certificates (DNS SANs, IP SANs, etc.). If I’m reading the code correctly this would require exposing a new configuration value (maybe ssl.verification.hostname?) and using it in rd_kafka_transport_ssl_set_endpoint_id() if it’s set. I’m skipping the checklist since this is an enhancement request, but I’m happy to provide more context for my use case if necessary.

Checklist
=========

Please provide the following information:

 - [x] librdkafka version (release number or git tag): `<REPLACE with e.g., v0.10.5 or a git sha. NOT "latest" or "current">`
 - [ ] Apache Kafka version: `<REPLACE with e.g., 0.10.2.3>`
 - [ ] librdkafka client configuration: `<REPLACE with e.g., message.timeout.ms=123, auto.reset.offset=earliest, ..>`
 - [ ] Operating system: `<REPLACE with e.g., Centos 5 (x64)>`
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
If it was a configuration property it would be applied to all broker connections, is that useful?

**Comment by crivera-fastly** [comment]
Yes. My use case involves a set of brokers running in the cloud with some clients that connect from within the vpc and some clients that connect from outside of the vpc. The internal and external clients use different broker hostnames. Getting both cases to work currently requires adding SANs to the certs. The brokers (foo1.bar.com, foo2.bar.com, …) serve a cert for foo.bar.com. This setting would let me configure all clients to accept the foo.bar.com cert.

**Comment by crivera-fastly** [comment]
Thanks for the detailed response @KJTsanaktsidis. I saw the existing verify_cb, but I definitely don’t want to be on the hook for validating the entire cert chain. A callback that would allow the client to specify a per-host verification override would work for my case and it's definitely more flexible than my initial suggestion.

**Comment by edenhill** [comment]
Thanks @KJTsanaktsidis.

@crivera-fastly I think I'm mostly in favour of the configuration property approach at this point before there is a required use-case for a callback. A property makes the SNI override functionality immediately available in all derrived clients and tools without any extra work needed, while a callback interface falls under the ABI guarantees and needs to be future proof and also requires work to be supported in all the sibling clients, et.al.

**Comment by crivera-fastly** [comment]
@edenhill That makes sense.
